File: examples/cs/gmreq/gmreq.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */




/*! \file gmreq.cs
 * 
 * A C# example demonstrating the GMSEC request/reply functionality.
 * The associated example gmrpl_cs will provide the reply.
 * gmrpl_cs should be run before gmreq_cs
 * 
*/


namespace gmreq
{
	using GMSEC.API;
	using GMSEC.API.UTIL;
	using GMSEC.API.Example.Common;
	using System;


	class gmreq : GmsecExample
	{
		private Config     config;
		private Connection conn;


		public gmreq()
		{
		}


		private void PrintUsage()
		{
			Console.Out.WriteLine(
				"\nusage: gmreq_cs connectiontype=<middleware> " +
				"[ <parameter>=<value> ]\n" +
				"\n\tNote that the parameter 'connectiontype' is required. " + 
				"\n\tThe rest of other parameters are optional.\n" +
				"\n\tserver=<server name> " +
				"(required if middleware is not bolt/MB locally)" +
				"\n\tsubject=<subject name>" +
				"\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
				"\n\tloglevel=<log level>" +
				"\n\tcfgfile=<config_filename.xml>" +
				"\n\nFor more information, see API User's Guide\n");
		}


		private void CleanUp()
		{
			if (conn != null)
			{
				conn.Disconnect();

				Connection.Destroy(ref conn);
			}
		}


		private void Request(int msg_timeout_ms, string subject)
		{
			// Create request message
			Message request = new Message(subject, Message.MessageKind.REQUEST);

			// Add fields to request message
			request.AddField(new StringField("QUESTION", "Does the request/reply functionality still work?"));
			request.AddField(new StringField("COMPONENT", "GMREQ"));

            if (config.GetBooleanValue("ENCRYPT", false))
            {
				request.AddField(new BooleanField("SEC-ENCRYPT", true));
            }

			// Display XML representation of the request message
			Log.Info("Prepared Request Message:\n" + request.ToXML());

            // Issue request message
            Message reply = conn.Request(request, msg_timeout_ms, 0);

			if (reply == null)
			{
				Log.Warning("request timed out");
			}
			else
			{
				// Display XML representation of the reply message
				Log.Info("Received Reply Message:\n" + reply.ToXML());

				// Destroy Message
				conn.Release(ref reply);
			}
		}


		private void Run(string[] args)
		{
			config = new Config(args);

			Initialize(config);
			AddToConfigFromFile(config);

			if (IsOptionInValid(config, args, false))
			{
				PrintUsage();
				return;
			}

			try
			{
				// Output GMSEC API version
				Log.Info(Connection.GetAPIVersion());

				// Create the connection
				conn = Connection.Create(config);

				// Connect
				conn.Connect();

				// Get CommandLine Info
				string subject     = Get(config, "SUBJECT", "GMSEC.MISSION.SAT_ID.REQ.CMD.GMRPL");
				int msg_timeout_ms = Get(config, "MSG_TIMEOUT_MS", (int) GMSECWaitDefs.WAIT_FOREVER);

				// OutPut MiddlewareVersion
				Log.Info("Middleware Version = " + conn.GetLibraryVersion());

				// OutPut Info
				Log.Info("Using Subject '" + subject + "'");

				// Prepare request message and issue request
				Request(msg_timeout_ms, subject);
			}
			catch (GMSEC_Exception e)
			{
				Log.Error("Exception: " + e.ToString());
			}
			finally
			{
				CleanUp();
			}
		}


		static void Main(string[] args)
		{
			new gmreq().Run(args);
		}
	}
}

File: examples/cpp/gmreq/gmreq.cpp
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file gmreq.cpp
 * 
 * A C++ example demonstration of GMSEC request / reply functionality.
 * The associated example gmrpl will provide the reply.  
 * gmrpl should be run before gmreq.
 * 
 */

#include "../example.h"

#include <string>

using namespace gmsec::api;


class gmreq
{
public:
	gmreq(Config& config);
	~gmreq();
	bool run();

private:
	Config&     config;
	Connection* connection;

	void request(int iteration, const std::string& subject);
};


gmreq::gmreq(Config& c)
	: config(c),
	  connection(0)
{
	// Initialize Config
	example::initialize(c);
}


gmreq::~gmreq()
{
	if (connection)
	{
		connection->disconnect();

		Connection::destroy(connection);
	}

	Connection::shutdownAllMiddlewares();
}


bool gmreq::run()
{
	bool success = true;

	/* output GMSEC API version */
	GMSEC_INFO << Connection::getAPIVersion();

	try
	{
		//o Create the Connection
		connection = Connection::create(config);

		//o Connect
		connection->connect();

		//o Get information from the command line */
		std::string subject        = example::get(config, "SUBJECT", "GMSEC.MISSION.SAT_ID.REQ.CMD.GMRPL");
		int         msg_timeout_ms = example::get(config, "MSG_TIMEOUT_MS", GMSEC_WAIT_FOREVER);

		//o output connection middleware version
		GMSEC_INFO << connection->getLibraryVersion();

		//o Output information
		GMSEC_INFO << "Using subject '" << subject.c_str() << "'";

		request(msg_timeout_ms, subject);
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


void gmreq::request(int msg_timeout_ms, const std::string& subject)
{
	//o Create message
	Message request(subject.c_str(), Message::REQUEST);

	//o Add fields to message
	request.addField("QUESTION", "Does the request/reply functionality still work?");
	request.addField("COMPONENT", "GMREQ");

	if (config.getBooleanValue("ENCRYPT", false))
	{
		request.addField("SEC-ENCRYPT", true);
	}

	//o Display XML representation of request message
	GMSEC_INFO << "Requesting:\n" << request.toXML();

	//o Send Request Message
	Message* reply = connection->request(request, msg_timeout_ms);

	if (reply)
	{
		//o Got reply; display XML representation of reply message
		GMSEC_INFO << "Received reply:\n" << reply->toXML();

		//o Destroy the reply message
		connection->release(reply);
	}
	else
	{
		GMSEC_WARNING << "Reply message not receive after a timeout of " << msg_timeout_ms << " msec.";
	}
}


int main(int argc, char* argv[])
{
	Config config(argc, argv);

	example::addToConfigFromFile(config);

	if (example::isOptionInvalid(config, argc))
	{
		example::printUsage("gmreq");
		return -1;
	}

	gmreq(config).run();
}
